tests/src/PatternAssertion/PaginationPatternAssert.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\Tests\oe_bootstrap_theme\PatternAssertion;
+
+use Symfony\Component\DomCrawler\Crawler;
+
+/**
+ * Assertions for the pagination pattern.
+ */
+class PaginationPatternAssert extends BasePatternAssert {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getAssertions(string $variant): array {
+    return [
+      'links' => [
+        [$this, 'assertLinks'],
+      ],
+      'alignment' => [
+        [$this, 'assertAlignment'],
+      ],
+      'size' => [
+        [$this, 'assertSize'],
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function assertBaseElements(string $html, string $variant): void {
+    $this->assertCounts([
+      'nav' => 1,
+      'ul' => 1,
+    ], new Crawler($html));
+  }
+
+  /**
+   * Asserts the pager links.
+   *
+   * @param array[] $expected
+   *   Expected pager links.
+   *   Each link is an array, with the following keys:
+   *     - 'url': The link destination.
+   *     - 'label': The label, if exists.
+   *     - 'icon': The name of the icon.
+   *     - 'active': Whether the item has an 'active' class.
+   * @param \Symfony\Component\DomCrawler\Crawler $crawler
+   *   The crawler.
+   */
+  protected function assertLinks(array $expected, Crawler $crawler): void {
+    $expected_bcl_icon_url = '/' . \Drupal::service('extension.list.theme')->getPath('oe_bootstrap_theme') . '/assets/icons/bcl-default-icons.svg';
+
+    $actual_links_data = array_map(
+      function (\DOMElement $actual_li_element) use ($expected_bcl_icon_url) {
+        $li = new Crawler($actual_li_element);
+        $link = $li->filter('a');
+        $actual_link_data = [];
+        $actual_link_data['url'] = $link->attr('href');
+        $use = $li->filter('a > svg > use');
+        if ($use->count()) {
+          // This is an icon.
+          $icon_url = $use->attr('xlink:href');
+          [$actual_bcl_icon_url, $actual_link_data['icon']] = explode('#', $icon_url, 2) + [NULL];
+          $this->assertSame($expected_bcl_icon_url, $actual_bcl_icon_url);
```

Comment:
If we want what you call "full asserts", I think we need to get there gradually. So we would have to endure a situation or a phase where we have both.

My motivation here was to cover the most common fails with a group assert that provides detailed information for fixing the test, and then have additional single assertions for things that will fail less often, and that are harder to cover in the group assertion.

The group assert failure output was already helpful in fixing the datasets.

This said, I agree for consistency it may be better to go for single asserts all the way, even if that makes the failure output less helpful.